common/pkg/libartifact/store/reference.go

```diff
@@ -0,0 +1,100 @@
+//go:build !remote
+
+package store
+
+import (
+	"context"
+
+	"go.podman.io/image/v5/docker/reference"
+)
+
+type ArtifactReference struct {
+	reference.Named
+}
+
+// NewArtifactReference is a theoretical reference to an artifact.  It needs to be
+// a fully qualified oci reference except for tag, where we add
+// "latest" as the tag if tag is empty.  Valid references:
+//
+// quay.io/podman/machine-os:latest
+// quay.io/podman/machine-os
+// quay.io/podman/machine-os@sha256:916ede4b2b9012f91f63100f8ba82d07ed81bf8a55d23c1503285a22a9759a1e
+//
+// Note: Partial sha references and digests (IDs) are not allowed.
+func NewArtifactReference(input string) (ArtifactReference, error) {
+	ar := ArtifactReference{}
+	named, err := stringToNamed(input)
+	if err != nil {
+		return ArtifactReference{}, err
+	}
+	ar.Named = named
+	return ar, nil
+}
+
+func (ar ArtifactReference) IsDigested() bool {
+	_, isDigested := ar.Named.(reference.Digested)
+	return isDigested
+}
+
+type ArtifactStoreReference struct {
+	ArtifactFromStore *Artifact
+	IsDigested        bool
+	Ref               reference.Named
+}
+
+// NewArtifactStorageReference refers to an object already in the artifact store.  It
+// can be a name or a full or partial digest.  Conveniently, it also embeds the artifact
+// as part of its return.
+func NewArtifactStorageReference(nameOrDigest string, as *ArtifactStore) (ArtifactStoreReference, error) {
```

Comment:
I think it would be good to mention that these type of functions require the caller to hold a lock.